blog-posts/typescript-ast-type-checker.md

```diff
@@ -0,0 +1,591 @@
+
+```typescript
+ts.forEachChild(mySourceFile, node => {
+  if (ts.isTypeAliasDeclaration(node) && node.name.escapedText === "Output") {
+    // [...process that type]
+  }
+});
+```
+By finding that node we have found the exact place in the source file to ask the type checker for more information. We can do this with the method `checker.getTypeAtLocation` and passing in the node, in return we get an instance of `ts.Type` from the checker.
+
+```typescript
+// Don't get confused, the `name` is not a string but an object with many more information
+// That's why this works even for things that are named the same
+const outputType = checker.getTypeAtLocation(node.name);
+```
+
+This is it, we arrive in type checking land 🌈
+
+We can access every property of the given type through `outputType.getProperties()`.
+
+```typescript
+const [collectedValueProperty] = outputType.getProperties();
+/**
+ * `propertyType` will contain & reference everything
+ * we can know about the type `ImportantValue`
+ */
+const propertyType = checker.getTypeOfSymbolAtLocation(collectedValueProperty, node);
+const propertyTypeName = checker.typeToString(propertyType);
+
+// prints `collectedValue: ImportantValue`
+console.log(`${collectedValueProperty.name}: ${propertyTypeName}`)
+```
+
+On that level we only have one property `collectedValue: ImportantValue` in our original type definition, so we can save us one loop and simply extract the first element and name it `collectedValueProperty`. The value has the type `ts.Symbol` which is similar to `ts.Type` a value with added semantics compared to the AST-related `ts.Node`.
+
+We can use that symbol, to access the name of the variable and the actual name of the type. The type checker gives us the methods `getTypeOfSymbolAtLocation` and `typeToString` to do that and we can print the final result to the console.
+
+```typescript
+// prints `collectedValue: ImportantValue`
+console.log(`└── ${collectedValueProperty.name}: ${propertyTypeName}`)
+```
+What's left is to dive one level deeper to finally extract the types from the nested `ImportantValue`. This is basically *"rinse & repeat"* as you will see in the following code example. Instead of extracing the first element we use a for-loop though in order to find all properties.
+
+```typescript
+// remember we are now processing `ImportantValue` which is stored in `propertyType`
+for (const nestedProperty of propertyType.getProperties()) {
+  const nestedPropertyType = checker.getTypeOfSymbolAtLocation(nestedProperty, node);
+  const nestedPropertyTypeName = checker.typeToString(nestedPropertyType);
+  /** prints the following
+   ├── value1: string
+   ├── value2: number
+   ├── value3: Date
+   */
+  console.log(`  ├── ${nestedProperty.name}: ${nestedPropertyTypeName}`)
+}
+```
+
+<details>
+    <summary>Full Source Example</summary>
+
+    ````typescript
```

Comment:
One tick too much ;-)

Reply:
this is by purpose here, because there is a nested tick group and you can make the ticks arbitrarily long (so it's four tickers) to group nested ticks (with three ticks) 🤓 

````
A three ticks examples wrapped by four ticks to make it a code group:

```
works 👌
```
````